utils_cv/action_recognition/model.py

```diff
@@ -475,19 +488,190 @@ def evaluate(
         )
         return ret
 
+    def _predict(self, frames, transform):
+        """ Todo """
+        clip = torch.from_numpy(np.array(frames))
+        # Transform frames and append batch dim
+        sample = torch.unsqueeze(transform(clip), 0)
+        sample = sample.to(torch_device())
+        output = self.model(sample)
+        scores = nn.functional.softmax(output, dim=1).data.cpu().numpy()[0]
+        return scores
+
+    def _filter_labels(
+        self,
+        id_score_dict: dict,
+        labels: List[str],
+        threshold: float = 0.0,
+        target_labels: List[str] = None,
+        filter_labels: List[str] = None,
+    ) -> Dict[str, int]:
+        """ Todo """
+        # Show only interested actions (target_labels) with a confidence score >= threshold
+        result = {}
+        for i, s in id_score_dict.items():
+            label = labels[i]
+            if (
+                (s < threshold)
+                or (target_labels is not None and label not in target_labels)
+                or (filter_labels is not None and label in filter_labels)
+            ):
+                continue
+
+            if label in result:
+                result[label] += s
+            else:
+                result[label] = s
+
+        return result
+
+    def predict_frames(
+        self,
+        window: deque,
+        scores_cache: deque,
+        scores_sum: np.ndarray,
+        is_ready: list,
+        averaging_size: int,
+        score_threshold: float,
+        labels: List[str],
+        target_labels: List[str],
+        transforms: Compose,
+        update_println: Callable,
+    ) -> None:
+        """ Predicts frames """
+        t = time()
+        scores = self._predict(window, transforms)
+        dur = time() - t
+
+        # Averaging scores across clips (dense prediction)
+        scores_cache.append(scores)
+        scores_sum += scores
+
+        if len(scores_cache) == averaging_size:
+            scores_avg = scores_sum / averaging_size
+
+            top5_id_score_dict = {
+                i: scores_avg[i] for i in (-scores_avg).argpartition(4)[:5]
+            }
+            top5_label_score_dict = self._filter_labels(
+                top5_id_score_dict,
+                labels,
+                threshold=score_threshold,
+                target_labels=target_labels,
+            )
+            top5 = sorted(top5_label_score_dict.items(), key=lambda kv: -kv[1])
+
+            # fps and preds
+            println = (
+                f"{1 // dur} fps"
+                + "<p style='font-size:20px'>"
+                + "<br>".join([f"{k} ({v:.3f})" for k, v in top5])
+                + "</p>"
+            )
+
+            # Plot final results nicely
+            update_println(println)
+            scores_sum -= scores_cache.popleft()
+
+        # Inference done. Ready to run on the next frames.
+        window.popleft()
+        if is_ready:
+            is_ready[0] = True
+
+    def predict_video(
```

Comment:
Should we split this function into two:
- One that provides the current functionality including the IPython display in a notebook
- One that takes a video and returns the prediction scores, without any plotting/printing/etc (ie how it would be used during a call to a deployed API)

Reply:
`predict_frames` is the function that basically returns the scores.
`predict_video` handles the current functionality of processing the video and displaying the ipython disply in the notebook. 

I moved some of the model-scoring specific code into `predict_frames` to make this distinction clearer

Currently, `predict_video` called `predict_frames`